pkg/lib/operatorstatus/monitor.go

```diff
@@ -0,0 +1,221 @@
+package operatorstatus
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	configv1client "github.com/openshift/client-go/config/clientset/versioned/typed/config/v1"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/lib/operatorclient"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// Wait time before we probe next while checking whether cluster
+	// operator API is available.
+	defaultProbeInterval = 1 * time.Minute
+
+	// Default size of the notification channel.
+	defaultNotificationChannelSize = 64
+)
+
+// NewMonitor returns a new instance of Monitor that can be used to continuously
+// update a clusteroperator resource and an instance of Sender that can be used
+// to send update notifications to it.
+//
+// The name of the clusteroperator resource to monitor is specified in name.
+func NewMonitor(name string, log *logrus.Logger, opClient operatorclient.ClientInterface, configClient configv1client.ConfigV1Interface) (Monitor, Sender) {
+	logger := log.WithField("monitor", "clusteroperator")
+	discovery := opClient.KubernetesInterface().Discovery()
+	releaseVersion := os.Getenv("RELEASE_VERSION")
+	names := split(name)
+
+	logger.Infof("monitoring the following components %s", names)
+
+	monitor := &monitor{
+		logger:         logger,
+		reporter:       newReporter(releaseVersion),
+		writer:         NewWriter(discovery, configClient),
+		notificationCh: make(chan *NotificationContext, defaultNotificationChannelSize),
+		names:          names,
+	}
+
+	return monitor, monitor
+}
+
+// NotificationContext contains all necessary information related to a notification.
+type NotificationContext struct {
+	// Name of the clusteroperator resource associated with this CSV.
+	Name string
+
+	// Current is the CSV for which we have received notification.
+	Current *v1alpha1.ClusterServiceVersion
```

Comment:
I think this is probably exposing a bit too much detail from the lower level - we really just need enough information to determine if we're progressing, and to determine which ClusterOperator status to update. (i.e. think about how this will change when we change v1alpha1.ClusterServiceVersion to v1.Operator)

Reply:
Let me see if I can make it a little more abstract.

Reply:
I decoupled `monitor` from csv specifics, so the monitor now accepts the following instead of a csv specific structure -
```
// MutatorFunc accepts an existing status object and appropriately mutates it
// to reflect the observed states.
type MutatorFunc func(existing *configv1.ClusterOperatorStatus) (new *configv1.ClusterOperatorStatus)

// Mutate is a wrapper for MutatorFunc
func (m MutatorFunc) Mutate(existing *configv1.ClusterOperatorStatus) (new *configv1.ClusterOperatorStatus) {
	return m(existing)
}

// NotificationFunc wraps a notification event. it returns the name of the
// cluster operator object associated and a mutator function that will set the
// new status for the cluster operator object.
type NotificationFunc func() (name string, mutator MutatorFunc)

// Get is a wrapper for NotificationFunc.
func (n NotificationFunc) Get() (name string, mutator MutatorFunc) {
	return n()
}
```

This way we can extend this to `subscription` or any other resources. @ecordell let me know your thoughts.